app/src/main/java/ch/epfl/cs311/wanderwave/model/remote/FirebaseConnection.kt

```diff
@@ -87,29 +87,38 @@ abstract class FirebaseConnection<T, U> {
         .addOnSuccessListener { Log.d("Firestore", "DocumentSnapshot successfully updated!") }
   }
 
-  fun deleteItem(item: T) {
+  open fun deleteItem(item: T) {
     val itemId = getItemId(item)
     deleteItem(itemId)
   }
 
-  fun deleteItem(itemId: String) {
+  open fun deleteItem(itemId: String) {
     db.collection(collectionName)
         .document(itemId)
         .delete()
         .addOnFailureListener { e -> Log.e("Firestore", "Error deleting document: ", e) }
         .addOnSuccessListener { Log.d("Firestore", "DocumentSnapshot successfully deleted!") }
   }
 
-  fun getItem(item: T): Flow<T> = getItem(getItemId(item))
+  open fun getItem(itemId: String): Flow<T> =
+      getItem(itemId) { _: DocumentSnapshot, _: MutableStateFlow<T?> -> }
 
-  open fun getItem(itemId: String): Flow<T> {
+  open fun getItem(
+      itemId: String,
+      onSuccess: (DocumentSnapshot, MutableStateFlow<T?>) -> Unit =
+          { _: DocumentSnapshot, _: MutableStateFlow<T?> ->
+          }
+  ): Flow<T> {
     val dataFlow = MutableStateFlow<T?>(null)
     db.collection(collectionName)
         .document(itemId)
         .get()
         .addOnSuccessListener { document ->
           if (document != null && document.data != null) {
-            documentToItem(document)?.let { dataFlow.value = it }
+            documentToItem(document)?.let {
+              dataFlow.value = it
+              onSuccess(document, dataFlow)
+            }
           }
         }
         .addOnFailureListener { e -> Log.e("Firestore", "Error getting document: ", e) }
```

Comment:
I don't understand why there is an onSuccess parameter, you should just be able to collect the value from the flow right?

Also the error case should probably be propagated to the flow, so that it's not possible to deadlock waiting for a failed item forever. Might be easier with the `flow { }` syntax.

Reply:
It's not for the user, it's more the other firebase function. This is the best solution I found, to do this. I can explain it to you more, but if you want to understand you can check the implementations of getitem for the beacon and profile connections.